modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/api/BundleGroupExtractor.java

```diff
@@ -0,0 +1,190 @@
+
+    protected final Map<String, List<String>> mavenSubGroups = new HashMap<>();
+
+    protected final List<BundleGroup> roots = new ArrayList<>();
+
+    protected final Map<String, BundleGroup> groups = new LinkedHashMap<>();
+
+    public static final String VIRTUAL_BUNDLE_GROUP = BundleGroup.PREFIX + "org.nuxeo.misc";
+
+    protected static final List<String> BLACKLIST_SUFFIX = List.of("test", "tests");
+
+    protected static final List<String> BLACKLIST = List.of("org", "org.nuxeo", "org.nuxeo.ecm",
+            "org.nuxeo.ecm.platform", "com", "com.nuxeo");
+
+    public BundleGroupExtractor(Map<String, BundleInfo> bundles, String version) {
+        this.bundles = bundles;
+        this.version = version;
+        for (BundleInfo bundle : bundles.values()) {
+            registerBundle(bundle);
+        }
+        generateGroups(version);
+    }
+
+    public List<BundleGroup> getRoots() {
+        return Collections.unmodifiableList(
+                roots.stream().sorted(new NuxeoArtifactComparator()).collect(Collectors.toList()));
+    }
+
+    public Map<String, BundleGroup> getGroups() {
+        return Collections.unmodifiableMap(groups);
+    }
+
+    protected void registerBundle(BundleInfo bundle) {
+        String groupId = bundle.getGroupId();
+        if (groupId != null) {
+            groupId = BundleGroup.PREFIX + groupId;
+        }
+        String artifactId = bundle.getArtifactId();
+        if (groupId == null || artifactId == null) {
+            groupId = VIRTUAL_BUNDLE_GROUP;
+            ((BundleInfoImpl) bundle).setGroupId(groupId);
+        }
+        if (!mavenGroups.containsKey(groupId)) {
+            mavenGroups.put(groupId, new ArrayList<String>());
+        }
+        mavenGroups.get(groupId).add(bundle.getId());
+    }
+
+    protected void generateGroups(String version) {
+        // post process bundle groups
+        List<String> mvnGroupNames = new ArrayList<>();
+        mvnGroupNames.addAll(mavenGroups.keySet());
+
+        for (String mvnGroupName : mvnGroupNames) {
+            List<String> artifactIds = mavenGroups.get(mvnGroupName);
+            Collections.sort(artifactIds);
+            Set<String> subGroups = new LinkedHashSet<>();
+            for (String id : artifactIds) {
+                if (id.contains(".")) {
+                    String suffix = id.substring(id.lastIndexOf(".") + 1);
+                    if (BLACKLIST_SUFFIX.contains(suffix.toLowerCase())) {
+                        continue;
+                    }
+                    String grpName = id.substring(0, id.lastIndexOf("."));
+                    String grp = BundleGroup.PREFIX + grpName;
+                    if (grp.equals(mvnGroupName) || BLACKLIST.contains(grpName.toLowerCase())) {
+                        continue;
+                    }
+                    subGroups.add(grp);
+                }
+            }
+            if (subGroups.size() < 1) {
```

Comment:
This was `< 2` in the original code, is there really a need to change this and do a split if size is 1?

Reply:
I did the extraction in a separate step to show the differences indeed.

This change allows to extract a group for "grp:org.nuxeo.ecm.platform.restapi" (visible in unit tests) as well a proper restoration of the apidoc group in json legacy tests (with only 2 bundles, apidoc-core and apidoc-repo --> the group would not be extracted in this case).

Overall the "< 2" felt arbitrary to me. I kept 1 instead of 0 because it might not make sense to keep a group for only one bundle.